src/trex/molecule.py

```diff
@@ -82,3 +85,141 @@ def compute_consensus(sequences):
     bin_consens = np.argmax(matrix, axis=1)
     # convert maximum counts into consensus sequence
     return "".join(letters[bin_consens])
+
+
+def remove_odd_clone_ids(molecules: List[Molecule],
+                        min_bases_detected: int = 7) -> List[Molecule]:
+    """Discard CloneIDs with a less than min_bases_detected or a
+    single (repeated) base detected."""
+    def acceptable_clone_id(clone_id):
+        detected_clone_id = re.sub("[-0]", "", clone_id)
+        if len(detected_clone_id) <= min_bases_detected or len(set(detected_clone_id)) <= 1:
+            return False
+        else:
+            return True
+
+    new_molecules = [mol for mol in molecules if acceptable_clone_id(
+        mol.clone_id)]
+    return new_molecules
+
+
+def correct_clone_ids(
+    molecules: List[Molecule], max_hamming: int, min_overlap: int = 20
+) -> List[Molecule]:
+    """
+    Attempt to correct sequencing errors in the cloneID sequences of all molecules
+    """
+    # Obtain all cloneIDs (including those with '-' and '0')
+    clone_ids = [m.clone_id for m in molecules]
+
+    # Count the full-length cloneIDs
+    counts = Counter(clone_ids)
+
+    # Cluster them by Hamming distance
+    def is_similar(s, t):
+        # m = max_hamming
+        if "-" in s or "-" in t:
+            # Remove suffix and/or prefix where sequences do not overlap
+            s = s.lstrip("-")
+            t = t[-len(s) :]
+            s = s.rstrip("-")
+            if len(s) < min_overlap:
+                return False
+            t = t[: len(s)]
+            # TODO allowed Hamming distance should be reduced relative to the overlap length
+            # m = max_hamming * len(s) / len(original_length_of_s)
+        return hamming_distance(s, t) <= max_hamming
+
+    clusters = cluster_sequences(list(set(clone_ids)), is_similar=is_similar, k=7)
+
+    # Map non-singleton cloneIDs to a cluster representative
+    clone_id_map = dict()
+    for cluster in clusters:
+        if len(cluster) > 1:
+            # Pick most frequent cloneID as representative
+            representative = max(cluster, key=lambda bc: (counts[bc], bc))
+            for clone_id in cluster:
+                clone_id_map[clone_id] = representative
+
+    # Create a new list of molecules in which the cloneIDs have been replaced
+    # by their representatives
+    new_molecules = []
+    for molecule in molecules:
+        clone_id = clone_id_map.get(molecule.clone_id, molecule.clone_id)
+        new_molecules.append(molecule._replace(clone_id=clone_id))
+    return new_molecules
+
+
+def correct_clone_ids_per_cell(
```

Comment:
Hm, there’s quite a bit of duplicated code. Can you perhaps unify this and the preceding function into one that accepts a `correct_per_cell: bool` parameter?

Reply:
I was thinking about this for quite some time, but I couldn't decide on how to do this. I thought maybe I could do some recursion by calling the correct_clone_ids function for each cell, but I noticed that I modified a couple things:

1. I modified the is_similar function because it was not symmetric (is_similar(string_a, string_b) could be different from is_similar(string_b, string_a)). I believe the heuristics behind this was that cloneIDs are alphabetically sorted.
2. I modified is_similar to remove 0s. I do not fully understand the heurstics behind ignoring "-" and not "0", but I went for stripping both.
3. When voting for a representative cloneID, the most abundant was used. In some cases, incoplete versions (containing "-") won over complete versions, so I decided to rank them according to how complete the molecules were, and then assess which one was the most abundant.

As this changes were spread across several lines, I made a separate function. Maybe there is a better way of combining them, or we could discuss if some of these functions should be defined outside this particular function and used across the package. Do you have some suggestion on how to proceed?

Reply:
I moved the change of `is_similar` to #44. I think it was just buggy and should be fixed.